Don't warn for polymorphic belongsTo type keys

wecc opened the pull request:
This is not pretty but it was the least ugly thing I could come up with. This also allows for regular attributes ending in "Type", like:

```javascript
App.Thing = DS.Model.extend({
  fakeType: DS.attr('string')
});
```

Fixes #2599

sly7-7 commented:
@wecc what do you think about extracting say, a `filterUnusedKeys` function and drying this thing

```
return filter(Ember.keys(data), function(key) {
  if (key.match(/Type$/)) {
    var relationship = get(type, 'relationshipsByName').get(key.replace(/Type$/, ''));
    if (relationship) {
      return relationship.kind !== 'belongsTo' && relationship.options.polymorphic !== true;
    }
  }
  return !get(type, 'fields').has(key) && key !== 'id' && key !== 'links';
})
```

Sounds possible ?

wecc commented:
@sly7-7 I would love to get away from the inline ugliness. The reason it's done like this is for defeatureify to be able to strip the whole `Ember.warn()` statement. I think we tried to do som `runInDebug()` voodoo in the beginning but defeatureify couldn't handle nested stuff.

sly7-7 commented:
@wecc I should have thought about that before asking... So at this point this looks good, thanks a lot for your work.

@igorT ok with merging this + #2678 ?

wecc commented:
After discussing with @igorT we simplified this for better readability and less code complexity. Also switched the warning to default-off and renamed the flag to `DS_WARN_ON_UNKNOWN_KEYS`.

Rebased.
